include/nbl/video/utilities/CArithmeticOps.h

```diff
@@ -0,0 +1,286 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_VIDEO_C_ARITHMETIC_OPS_H_INCLUDED_
+#define _NBL_VIDEO_C_ARITHMETIC_OPS_H_INCLUDED_
+
+
+#include "nbl/video/IPhysicalDevice.h"
+#include "nbl/video/utilities/IDescriptorSetCache.h"
+
+#include "nbl/builtin/hlsl/scan/declarations.hlsl"
+
+
+namespace nbl::video
+{
+class CArithmeticOps : public core::IReferenceCounted
+{
+    public:
+		// Only 4 byte wide data types supported due to need to trade the via shared memory,
+		// different combinations of data type and operator have different identity elements.
+		// `EDT_INT` and `EO_MIN` will have `INT_MAX` as identity, while `EDT_UINT` would have `UINT_MAX`  
+		enum E_DATA_TYPE : uint8_t
+		{
+			EDT_UINT=0u,
+			EDT_INT,
+			EDT_FLOAT,
+			EDT_COUNT
+		};
+		enum E_OPERATOR : uint8_t
+		{
+			EO_AND = 0u,
+			EO_XOR,
+			EO_OR,
+			EO_ADD,
+			EO_MUL,
+			EO_MIN,
+			EO_MAX,
+			EO_COUNT
+		};
+
+		// This struct is only for managing where to store intermediate results of the scans
+		struct Parameters : nbl::hlsl::scan::Parameters_t // this struct and its methods are also available in HLSL
+		{
+			static inline constexpr uint32_t MaxLevels = NBL_BUILTIN_MAX_LEVELS;
+
+			Parameters()
+			{
+				std::fill_n(lastElement,MaxLevels/2+1,0u);
+				std::fill_n(temporaryStorageOffset,MaxLevels/2,0u);
+			}
+			// build the constant tables for each level given the number of elements to scan and workgroupSize
+			Parameters(const uint32_t _elementCount, const uint32_t workgroupSize) : Parameters()
+			{
+				assert(_elementCount!=0u && "Input element count can't be 0!");
+				const auto maxReductionLog2 = hlsl::findMSB(workgroupSize)*(MaxLevels/2u+1u);
+				assert(maxReductionLog2>=32u||((_elementCount-1u)>>maxReductionLog2)==0u && "Can't scan this many elements with such small workgroups!");
+
+				lastElement[0u] = _elementCount-1u;
+				for (topLevel=0u; lastElement[topLevel]>=workgroupSize;)
+					temporaryStorageOffset[topLevel-1u] = lastElement[++topLevel] = lastElement[topLevel]/workgroupSize;
+				
+				std::exclusive_scan(temporaryStorageOffset,temporaryStorageOffset+sizeof(temporaryStorageOffset)/sizeof(uint32_t),temporaryStorageOffset,0u);
+			}
+            // given already computed tables of lastElement indices per level, number of levels, and storage offsets, tell us total auxillary buffer size needed
+			inline uint32_t getScratchSize(uint32_t ssboAlignment=256u)
+			{
+				uint32_t uint_count = 1u; // workgroup enumerator
+				uint_count += temporaryStorageOffset[MaxLevels/2u-1u]; // last scratch offset
+				uint_count += lastElement[topLevel]+1u; // and its size
+				return core::roundUp<uint32_t>(uint_count*sizeof(uint32_t),ssboAlignment);
+			}
+		};
```

Comment:
this should not deviate from `nbl::hlsl::scan::Parameters_t` at all, if some of the constructors/methods you need on CPU are not there, move them there